src/ansi.rs

```diff
@@ -0,0 +1,1849 @@
+
+/// Parse colors in XParseColor format.
+fn xparse_color(color: &[u8]) -> Option<Rgb> {
+    if !color.is_empty() && color[0] == b'#' {
+        parse_legacy_color(&color[1..])
+    } else if color.len() >= 4 && &color[..4] == b"rgb:" {
+        parse_rgb_color(&color[4..])
+    } else {
+        None
+    }
+}
+
+/// Parse colors in `rgb:r(rrr)/g(ggg)/b(bbb)` format.
+fn parse_rgb_color(color: &[u8]) -> Option<Rgb> {
+    let mut colors = str::from_utf8(color).ok()?.split('/');
+    let red = colors.next()?;
+    let green = colors.next()?;
+    let blue = colors.next()?;
+
+    if colors.next().is_some() {
+        return None;
+    }
+
+    // Scale values instead of filling with `0`s.
+    let scale = |input: &str| {
+        if input.len() > 4 {
+            None
+        } else {
+            let max = u32::pow(16, input.len() as u32) - 1;
+            let value = u32::from_str_radix(input, 16).ok()?;
+            Some((255 * value / max) as u8)
+        }
+    };
+
+    Some(Rgb { r: scale(red)?, g: scale(green)?, b: scale(blue)? })
+}
+
+/// Parse colors in `#r(rrr)g(ggg)b(bbb)` format.
+fn parse_legacy_color(color: &[u8]) -> Option<Rgb> {
+    let item_len = color.len() / 3;
+
+    // Truncate/Fill to two byte precision.
+    let color_from_slice = |slice: &[u8]| {
+        let col = usize::from_str_radix(str::from_utf8(slice).ok()?, 16).ok()? << 4;
+        Some((col >> (4 * slice.len().saturating_sub(1))) as u8)
+    };
+
+    Some(Rgb {
+        r: color_from_slice(&color[0..item_len])?,
+        g: color_from_slice(&color[item_len..item_len * 2])?,
+        b: color_from_slice(&color[item_len * 2..])?,
+    })
+}
+
+fn parse_number(input: &[u8]) -> Option<u8> {
+    if input.is_empty() {
+        return None;
+    }
+    let mut num: u8 = 0;
+    for c in input {
+        let c = *c as char;
+        if let Some(digit) = c.to_digit(10) {
+            num = match num.checked_mul(10).and_then(|v| v.checked_add(digit as u8)) {
+                Some(v) => v,
+                None => return None,
+            }
+        } else {
+            return None;
+        }
+    }
+    Some(num)
+}
+
+/// Internal state for VTE processor.
+struct ProcessorState<T: TimeProvider> {
+    /// Last processed character for repetition.
+    preceding_char: Option<char>,
+
+    /// DCS sequence waiting for termination.
+    #[cfg(feature = "alloc")]
+    dcs: Option<Dcs>,
+
+    /// State for synchronized terminal updates.
+    #[cfg(feature = "alloc")]
+    sync_state: SyncState<T>,
+
+    _time_provider: PhantomData<T>,
+}
+
+/// Current timestamp provider
+pub trait TimeProvider {
+    /// Get current timestamp from unix epoch in milliseconds
+    fn now(&self) -> u128;
+}
+
+/// Time provider which always returns 0
+#[derive(Default)]
+pub struct NullTimeProvider {}
+
+impl TimeProvider for NullTimeProvider {
+    fn now(&self) -> u128 {
+        0
+    }
+}
+
+#[derive(Debug)]
+#[cfg(feature = "alloc")]
```

Comment:
```suggestion
#[cfg(feature = "alloc")]
#[derive(Debug)]
```

Logically I think it makes way more sense to have the `cfg` attribute before any other macro. So I think it would be good to change this everywhere so that the scope under which the struct is defined is the first thing people read.